### PR TITLE
Optimize the vertex formats of meshes for less input buffer bindings

### PR DESCRIPTION
**Pull request by samcake**
This pr does 2 things:
- It interleaves the attributes in the vertex buffer of the Mesh to optimize the Vertex format so there are only 1 or 3 (for blend shaped) buffers bound as Input Buffer instead of 7! currently.
- it uses directly the gl Object ID for buffers during the Draw pass of the Batch execution because by then we are guaranteed that the gl id are valid and the Buffers have been synchronized correctly. This is shaving a bit of extra checks during the batch execution.

## TEST PLAN ##
- On every OS and GPU platforms check that:
   meshes render correctly
   skinned meshes are rendering correctly (avatars)
   blend shaped meshes are rendering correctly (avatar's Face while talking to some one with a humanoid avatar)

- Verify that in HMD there is no lagging of the rendering vs the head movment

- Do the regular auto test pass about  engine/render/geometry:
https://github.com/highfidelity/hifi_tests/blob/master/tests/engine/render/geometry?raw=true
and content/entity/model:
https://github.com/highfidelity/hifi_tests/blob/master/tests/content/entity/model?raw=true

- Observed performances in heavy scenes (engine-dev with 600 bots, dev-avatarisland) of the time spent in rendering (REnder engine, Batch or GPU) should be equal are better (less time) than in current master

### CODE REVIEW

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12598/HighFidelity-Beta-PR13922-0fe13ea.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12598/HighFidelity-Beta-Interface-PR13922-0fe13ea.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12598/HighFidelity-Beta-PR13922-0fe13ea.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12598/HighFidelity-Beta-Interface-PR13922-0fe13ea.exe

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12620/HighFidelity-Beta-PR13922-761e8a3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12620/HighFidelity-Beta-Interface-PR13922-761e8a3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12676/HighFidelity-Beta-PR13922-f5843c2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12676/HighFidelity-Beta-Interface-PR13922-f5843c2.dmg

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12686/HighFidelity-Beta-PR13922-15f54ac.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12686/HighFidelity-Beta-Interface-PR13922-15f54ac.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12715/HighFidelity-Beta-PR13922-e428248.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12715/HighFidelity-Beta-Interface-PR13922-e428248.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12715/HighFidelity-Beta-PR13922-e428248.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12715/HighFidelity-Beta-Interface-PR13922-e428248.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12725/HighFidelity-Beta-PR13922-6544e9f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12725/HighFidelity-Beta-PR13922-6544e9f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12725/HighFidelity-Beta-Interface-PR13922-6544e9f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12725/HighFidelity-Beta-PR13922-6544e9f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12725/HighFidelity-Beta-Interface-PR13922-6544e9f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12738/HighFidelity-Beta-PR13922-7349562.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12738/HighFidelity-Beta-Interface-PR13922-7349562.dmg

File: libraries/gpu-gl-common/src/gpu/gl/GLBuffer.h [hunk 1]
```diff
@@ -49,13 +49,23 @@ class GLBuffer : public GLObject<Buffer> {
...
+    }
+
     const GLuint& _buffer { _id };
     const GLuint _size;
     const Stamp _stamp;
 
     ~GLBuffer();
 
-    virtual void transfer() = 0;
+    virtual void transfer() {};
```
Comment (jherico) [inline libraries/gpu-gl-common/src/gpu/gl/GLBuffer.h]: why this change?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12738/HighFidelity-Beta-PR13922-7349562.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12738/HighFidelity-Beta-Interface-PR13922-7349562.exe

File: libraries/gpu-gl-common/src/gpu/gl/GLBuffer.h (hunk 1 above)
Comment (samcake) [reply libraries/gpu-gl-common/src/gpu/gl/GLBuffer.h]: i don't know

File: libraries/fbx/src/FBXReader_Mesh.cpp [hunk 2]
```diff
@@ -705,69 +719,167 @@ void FBXReader::buildModelMesh(FBXMesh& extractedMesh, const QString& url) {
...
+            for (int32_t i = 0; i < numIndices; ++i) {
+                assert(fbxMesh.clusterIndices[i] <= UINT8_MAX);
+                clusterIndices[i] = (uint8_t)(fbxMesh.clusterIndices[i]);
+            }
+            vertBuffer->setSubData(clusterIndicesOffset, clusterIndicesSize, (const gpu::Byte*) clusterIndices.constData());
         }
-        attribBuffer->setSubData(clusterIndicesOffset, clusterIndicesSize, (const gpu::Byte*) clusterIndices.constData());
-    } else {
-        attribBuffer->setSubData(clusterIndicesOffset, clusterIndicesSize, (const gpu::Byte*) fbxMesh.clusterIndices.constData());
+        else {
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXReader_Mesh.cpp]: :)

File: libraries/fbx/src/FBXReader_Mesh.cpp [hunk 3]
```diff
@@ -705,69 +719,167 @@ void FBXReader::buildModelMesh(FBXMesh& extractedMesh, const QString& url) {
...
+    gpu::BufferPointer attribBuffer;
+    int totalAttribBufferSize = totalVertsSize;
+    gpu::uint8 posChannel = 0;
+    gpu::uint8 tangentChannel = posChannel;
+    gpu::uint8 attribChannel = posChannel;
+    bool interleavePositions = true;
+    bool interleaveNormalsTangents = true;
+
+    // If has blend shapes allocate and assign buffers for pos and tangents now
+    hasBlendShapes = true;
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXReader_Mesh.cpp]: ?

File: libraries/fbx/src/FBXReader_Mesh.cpp [hunk 4]
```diff
@@ -705,69 +719,167 @@ void FBXReader::buildModelMesh(FBXMesh& extractedMesh, const QString& url) {
...
+    if (hasBlendShapes) {
+        auto posBuffer = std::make_shared<gpu::Buffer>();
+        posBuffer->setData(positionsSize, (const gpu::Byte*) vertBuffer->getData() + positionsOffset);
+        vertexBufferStream->addBuffer(posBuffer, 0, positionElement.getSize());
+
+        auto tangentBuffer = std::make_shared<gpu::Buffer>();
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXReader_Mesh.cpp]: normalsAndTangentsBuffer

File: libraries/fbx/src/FBXReader_Mesh.cpp [hunk 5]
```diff
@@ -705,69 +719,167 @@ void FBXReader::buildModelMesh(FBXMesh& extractedMesh, const QString& url) {
...
-                            graphics::BufferView(attribBuffer, texCoords1Offset, texCoords1Size,
-                            gpu::Element(gpu::VEC2, gpu::HALF, gpu::UV)));
-    } else if (texCoordsSize) {
-        mesh->addAttribute(gpu::Stream::TEXCOORD1,
-                           graphics::BufferView(attribBuffer, texCoordsOffset, texCoordsSize,
-                           gpu::Element(gpu::VEC2, gpu::HALF, gpu::UV)));
+        vertexFormat->setAttribute(gpu::Stream::TEXCOORD1, attribChannel, texCoordsElement, bufOffset);
+        bufOffset += texCoordsElement.getSize();
+    }
+    else if (texCoordsSize) {
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXReader_Mesh.cpp]: `} {`

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12774/HighFidelity-Beta-PR13922-d979d02.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12774/HighFidelity-Beta-Interface-PR13922-d979d02.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12774/HighFidelity-Beta-PR13922-d979d02.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12774/HighFidelity-Beta-Interface-PR13922-d979d02.exe

File: libraries/fbx/src/FBXReader_Mesh.cpp (hunk 3 above)
Comment (samcake) [reply libraries/fbx/src/FBXReader_Mesh.cpp]: oups....

File: libraries/fbx/src/FBXReader_Mesh.cpp (hunk 4 above)
Comment (samcake) [reply libraries/fbx/src/FBXReader_Mesh.cpp]: ok

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12779/HighFidelity-Beta-PR13922-fa72910.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12779/HighFidelity-Beta-Interface-PR13922-fa72910.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12779/HighFidelity-Beta-PR13922-fa72910.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12779/HighFidelity-Beta-Interface-PR13922-fa72910.exe

**Comment by qaAdamSmith82** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12785/HighFidelity-Beta-PR13922-81a5429.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12785/HighFidelity-Beta-Interface-PR13922-81a5429.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12785/HighFidelity-Beta-PR13922-81a5429.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12785/HighFidelity-Beta-Interface-PR13922-81a5429.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12796/HighFidelity-Beta-PR13922-e8cfac2.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12799/HighFidelity-Beta-PR13922-5140999.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12799/HighFidelity-Beta-PR13922-5140999.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12799/HighFidelity-Beta-Interface-PR13922-5140999.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12799/HighFidelity-Beta-PR13922-5140999.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12799/HighFidelity-Beta-Interface-PR13922-5140999.exe

**Comment by qaAdamSmith82** [comment]
Stopped Testing

**Comment by qaAdamSmith82** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12817/HighFidelity-Beta-PR13922-fed377d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12817/HighFidelity-Beta-PR13922-fed377d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12817/HighFidelity-Beta-Interface-PR13922-fed377d.dmg

**Comment by qaAdamSmith82** [comment]
FAIL @conklin94122 

Numbers across testable platforms have increased.


master
amd13922
masteramd

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12817/HighFidelity-Beta-PR13922-fed377d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12817/HighFidelity-Beta-Interface-PR13922-fed377d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12824/HighFidelity-Beta-PR13922-082d47d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12824/HighFidelity-Beta-PR13922-082d47d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12824/HighFidelity-Beta-Interface-PR13922-082d47d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12824/HighFidelity-Beta-PR13922-082d47d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12824/HighFidelity-Beta-Interface-PR13922-082d47d.exe

**Comment by qaAdamSmith82** [comment]
PASS @conklin94122 Times are now equivalent, or extremely close to equivalent

master
pr
